replicate-to-sql-with-generic-events/README.md

```diff
@@ -0,0 +1,190 @@
+
+```javascript
+async function loadData() {
+  const documents = await getDocuments();
+
+  if (documents.length > 0) {
+    const response = await kuzzle.document.mCreate(indexName, collectionName, documents);
+    console.log(`Created ${response.successes.length} documents`);
+  } else {
+    throw new Error('No documents to insert');
+  }
+}
+```
+
+We then complete the script by
+
+1. connecting to the running kuzzle instance.
+1. creating the index/collection that will hold our data
+1. loading data to Kuzzle
+
+```javascript
+async function run() {
+  try {
+    await kuzzle.connect();
+    await createIndexIfNotExists(kuzzle, indexName);
+    await createCollectionIfNotExists(kuzzle, indexName, collectionName);
+    await loadData();
+  } finally {
+    kuzzle.disconnect();
+  }
+}
+```
+
+Looking at the [plugin file](lib/index.js), we can observe how **Generic Events** are being catch by the Plugin.
+
+```javascript
+class CorePlugin {
+  constructor() {
+    this.context = null;
+    this.config = {};
+    this.pipes = {
+      'generic:document:afterWrite': 'afterWrite',
+      'generic:document:afterDelete': 'afterDelete'
+    };
+  }
+}
+```
+
+By declaring `this.pipes` inside the constructor of the plugin we can catch events emitted by the core of kuzzle. Here, we will be listening to
+
+1. `generic:document:afterWrite` an event emitted right after documents have been written.
+1. `generic:document:afterDelete` an event emitted right after documents have been deleted.
+
+```javascript
+  async afterWrite(documents = []) {
+    const promises = documents.map(doc => this.pg.insert(this.getProperties(doc)));
+    await Promise.all(promises);
+
+    return documents;
+  }
+
+  async afterDelete(documents = []) {
+    const promises = documents.map(doc => this.pg.delete(doc._id));
+    await Promise.all(promises);
+
+    return documents;
+  }
+```
+
+More generic and non-generic events can be used: [Kuzzle events documentation](https://docs.kuzzle.io/core/2/plugins/guides/events/intro).
+
+We used the well-known [node-postgres](https://node-postgres.com/) client to interface our plugin with a postgres database.
+
+```javascript
+const { Pool } = require('pg');
+
+function createPool(config) {
+  const pool = new Pool(config);
+}
+```
+
+We will use the [Pool](https://node-postgres.com/api/pool) constructor to instanciate the Postgres driver within the kuzzle plugin.
+
+The pool acquires a client from the pool. If the pool is 'full' and all clients are currently checked out, this will wait in a FIFO queue until a client becomes available by it being released back to the pool.
+
+To make a correct insert inside the database, we will use a multiline insert. Since generic events will send a set of records as a payload, it is wise to insert them together rather than one by one.
```

Comment:
Using multiple INSERT queries is not incorrect, it's just slow.

```suggestion
For better performances, since generic events provide a set of records as their payload, we'll send them to Postgres using a single multilines INSERT.
```